ecc/secp256k1/ecdsa/ecdsa.go

```diff
@@ -110,6 +112,48 @@ func HashToInt(hash []byte) *big.Int {
 	return ret
 }
 
+// RecoverP recovers the value P (prover commitment) when creating a signature.
+// It uses the recovery information v and part of the decomposed signature r. It
+// is used internally for recovering the public key.
+func RecoverP(v uint, r *big.Int) (*secp256k1.G1Affine, error) {
+	x := new(big.Int).Set(r)
+	// if x is r or r+N
+	xChoice := (v & 15) >> 1
+	// if y is y or -y
+	yChoice := v & 1
+	// decompose limbs into big.Int value
+	// conditional +n based on xChoice
+	kn := big.NewInt(int64(xChoice))
+	kn.Mul(kn, fr.Modulus())
+	x.Add(x, kn)
+	// y^2 = x^3+ax+b
+	a, b := secp256k1.CurveCoefficients()
+	y := new(big.Int).Exp(x, big.NewInt(3), fp.Modulus())
+	if !a.IsZero() {
+		y.Add(y, new(big.Int).Mul(a.BigInt(new(big.Int)), x))
+	}
+	y.Add(y, b.BigInt(new(big.Int)))
+	y.Mod(y, fp.Modulus())
+	// y = sqrt(y^2)
+	if y.ModSqrt(y, fp.Modulus()) == nil {
```

Comment:
perf-wise it should be better to use `Sqrt` of `fp` package

Reply:
Will change.

Reply:
Well it's not really necessary unless perf is critical. This also applies to `ModInverse` used in `Sign` and `Verify` (before this PR). Maybe we can keep as is and do a tiny PR for both (if benchmark suggest it).